tasks/consensus/task.go

```diff
@@ -0,0 +1,63 @@
+package consensus
+
+import (
+	"context"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/sentinel-visor/model"
+	"github.com/filecoin-project/sentinel-visor/model/chain"
+	visormodel "github.com/filecoin-project/sentinel-visor/model/visor"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/label"
+)
+
+type Task struct {
+}
+
+func NewTask() *Task {
+	return &Task{}
+}
+
+func (t *Task) ProcessTipSets(ctx context.Context, child, parent *types.TipSet) (model.Persistable, visormodel.ProcessingReportList, error) {
+	_, span := global.Tracer("").Start(ctx, "ProcessTipSets")
+	if span.IsRecording() {
+		span.SetAttributes(label.String("child", child.String()), label.Int64("height", int64(child.Height())))
+		span.SetAttributes(label.String("parent", parent.String()), label.Int64("height", int64(parent.Height())))
+	}
+	defer span.End()
+
+	var pl chain.ChainConsensusList
+	var rp visormodel.ProcessingReportList
+	for epoch := child.Height(); epoch > parent.Height(); epoch-- {
+		if child.Height() == epoch {
+			pl = append(pl, &chain.ChainConsensus{
+				Height:          int64(epoch),
+				ParentStateRoot: child.ParentState().String(),
+				ParentTipSet:    child.Parents().String(),
+				TipSet:          child.Key().String(),
+			})
+			rp = append(rp, &visormodel.ProcessingReport{
+				Height:    int64(epoch),
+				StateRoot: child.ParentState().String(),
+			})
+		} else {
+			// null round no tipset
+			pl = append(pl, &chain.ChainConsensus{
+				Height:          int64(epoch),
+				ParentStateRoot: parent.ParentState().String(),
```

Comment:
Why parent.ParentState and not child.ParentState?

I thought child.ParentState was the state root after the parent tipset has been applied. It's the basis for the state that the child messages will be working with.

On that basis the null rounds should also use the state root after the parent tipset has been applied, i.e. child.ParentState.

The result would be:

```
height | parent_state_root          | note
--------------------------------------------------------------------------------
   H-1 |         www                | older tipset applied and creates new stateroot xxx
   H   |         xxx                | parent tipset applied and creates new stateroot yyy
   H+1 |         yyy                | null round, if it hadn't been null then yyy would have been used
   H+2 |         yyy                | null round
   H+3 |         yyy                | child tipset applied on yyy
```

Reply:
I am stating that the parent state root of a null round is equal to the parent state of the last non-null round epoch.

### Consensus View Calibration Net
|height|parent_state_root                                             |parent_tip_set |tip_set|
|-|-|-|-|
|1     |bafy2bzacebqhvfgoflvwukha3hjukmv7eceaxbxqrgzzkdu7kh6hr2z5vugfg|{bafy2bzacecz3trtejxtzix4f4eebs7dekm6snfsmvffiqz2rfx7iwgsgtieq4}|{bafy2bzacebbzkct2j2kssnnunxxblza3jbiot3sokyslltgk7ovdh2osz52sy}|
|2     |bafy2bzacebqhvfgoflvwukha3hjukmv7eceaxbxqrgzzkdu7kh6hr2z5vugfg|{bafy2bzacecz3trtejxtzix4f4eebs7dekm6snfsmvffiqz2rfx7iwgsgtieq4}|NULL|
|3     |bafy2bzacebqhvfgoflvwukha3hjukmv7eceaxbxqrgzzkdu7kh6hr2z5vugfg|{bafy2bzacecz3trtejxtzix4f4eebs7dekm6snfsmvffiqz2rfx7iwgsgtieq4}|NULL|
|4     |bafy2bzacedzb5swfiji7fxchfde7lbpbcqrlipd6s7jrqyyqyqyjt2gbdozxy|{bafy2bzacebbzkct2j2kssnnunxxblza3jbiot3sokyslltgk7ovdh2osz52sy}|{bafy2bzacedsredymetuudfpqrav4lomm5cmip6jysm2pinm34bqopkd4luqmi}|
|5     |bafy2bzacedhn62mqsspltflvocr3v3vye22ksmxooi3psys3ujcynw4k6vkae|{bafy2bzacedsredymetuudfpqrav4lomm5cmip6jysm2pinm34bqopkd4luqmi}|{bafy2bzaceaphl5mowqzr3hk6f76elh55g7hcxumbhrv2hr55e2qph7xojbjqi}|
|6     |bafy2bzacedhn62mqsspltflvocr3v3vye22ksmxooi3psys3ujcynw4k6vkae|{bafy2bzacedsredymetuudfpqrav4lomm5cmip6jysm2pinm34bqopkd4luqmi}|NULL|
|7     |bafy2bzacedw64lvrikpq6rgnmeffvl7zcgd7g7pkjmpqvrgapc3i6uo5q56gy|{bafy2bzaceaphl5mowqzr3hk6f76elh55g7hcxumbhrv2hr55e2qph7xojbjqi}|{bafy2bzacedfjhau5khoovyxwy43yddn7dahjekuq63snbisecxwntppnadgoi}|
|8     |bafy2bzacebhj3bdsayvvz4wvlz4sfaowotbryjzfrdmnwnzx7dnr5pgzfhyge|{bafy2bzacedfjhau5khoovyxwy43yddn7dahjekuq63snbisecxwntppnadgoi}|{bafy2bzaceddfcedna7rsv7mz3wc37aqcw7l6altgykwvbt5pa5jr24sypzrym}|
|9     |bafy2bzaceb27xrflyoupzn3xe4wuive34pj2bx45ffozyetjvrexafthccgvs|{bafy2bzaceddfcedna7rsv7mz3wc37aqcw7l6altgykwvbt5pa5jr24sypzrym}|{bafy2bzacedemnvd6lsztc7qvcmkmnch2y3d36cg4efkn5arcymebwgk6zleig}|
|10    |bafy2bzaceaivwetm6x5quwhh32fixdfvkqjubw4gch3doq6ml4wlnpbn5lhpa|{bafy2bzacedemnvd6lsztc7qvcmkmnch2y3d36cg4efkn5arcymebwgk6zleig}|{bafy2bzaced3453rgut5x4ei5isg3mtobcpjfniruwlemwwld7b6s2p4mx4ik2}|
|11    |bafy2bzacebklafbuxj7z6emsxhpcbkkw4njnzg6fevvdthh6q5ajtyfriagis|{bafy2bzaced3453rgut5x4ei5isg3mtobcpjfniruwlemwwld7b6s2p4mx4ik2}|{bafy2bzacearr4gjbmwhsjwac4qt5unokoaen3tay6rcqrmrsupolfrsiuhgxq}|

### Block Headers Calibration Net
|height|parent_state_root|                                              cid                                                           |parent_weight|
|------|-----------------|                                             ---------------------------------------------------------------|-------------|
|0     |bafy2bzacebqhvfgoflvwukha3hjukmv7eceaxbxqrgzzkdu7kh6hr2z5vugfg| bafy2bzacecz3trtejxtzix4f4eebs7dekm6snfsmvffiqz2rfx7iwgsgtieq4|0            |
|1     |bafy2bzacebqhvfgoflvwukha3hjukmv7eceaxbxqrgzzkdu7kh6hr2z5vugfg| bafy2bzacebbzkct2j2kssnnunxxblza3jbiot3sokyslltgk7ovdh2osz52sy|12544        |
|4     |bafy2bzacedzb5swfiji7fxchfde7lbpbcqrlipd6s7jrqyyqyqyjt2gbdozxy| bafy2bzacedsredymetuudfpqrav4lomm5cmip6jysm2pinm34bqopkd4luqmi|26342        |
|5     |bafy2bzacedhn62mqsspltflvocr3v3vye22ksmxooi3psys3ujcynw4k6vkae| bafy2bzaceaphl5mowqzr3hk6f76elh55g7hcxumbhrv2hr55e2qph7xojbjqi|40140        |
|7     |bafy2bzacedw64lvrikpq6rgnmeffvl7zcgd7g7pkjmpqvrgapc3i6uo5q56gy| bafy2bzacedfjhau5khoovyxwy43yddn7dahjekuq63snbisecxwntppnadgoi|55192        |
|8     |bafy2bzacebhj3bdsayvvz4wvlz4sfaowotbryjzfrdmnwnzx7dnr5pgzfhyge| bafy2bzaceddfcedna7rsv7mz3wc37aqcw7l6altgykwvbt5pa5jr24sypzrym|71499        |
|9     |bafy2bzaceb27xrflyoupzn3xe4wuive34pj2bx45ffozyetjvrexafthccgvs| bafy2bzacedemnvd6lsztc7qvcmkmnch2y3d36cg4efkn5arcymebwgk6zleig|85297        |
|10    |bafy2bzaceaivwetm6x5quwhh32fixdfvkqjubw4gch3doq6ml4wlnpbn5lhpa| bafy2bzaced3453rgut5x4ei5isg3mtobcpjfniruwlemwwld7b6s2p4mx4ik2|100349       |

Reply:
At epoch 2 the parent state root has to be different because the tipset at epoch 1 has been applied. 

`bafy2bzacebbzkct2j2kssnnunxxblza3jbiot3sokyslltgk7ovdh2osz52sy` is the state root after the tipset at epoch 1 has been applied. If a tipset had been selected for epoch 2 then that is the parent state root that would have been used.

Reply:
> I thought child.ParentState was the state root after the parent tipset has been applied. It's the basis for the state that the child messages will be working with.

100% agree with this. child.ParentState == Apply(parent.Messages)

> At epoch 2 the parent state root has to be different because the tipset at epoch 1 has been applied.

This isn't true. Epoch 2 (& 3) technically doesn't exist. So the messages in tipset 1 haven't been applied and thus the stateroot cannot be known until epoch 4. Similarly, epoch 6 doesn't technically exist, meaning the messages in tipset 5 haven't been applied yet. 

Stated differently, the chain state can not change at null rounds.

Reply:
I suppose one could argue that state _does_ exist at epoch 2 & 3 since miners applied messages but no leaders were elected to mine a block(s), but that information doesn't make it on-chain until epoch 4 when someone(s) wins that round.

Reply:
Let's be compatible with what the Lotus API does

Reply:
Then this should stay as is (api: https://github.com/filecoin-project/lotus/blob/57bf5c21433df6afc9c14b33b182bfa53a9ad3b4/api/api_full.go#L111-L114)

Reply:
Why not https://github.com/filecoin-project/lotus/blob/57bf5c21433df6afc9c14b33b182bfa53a9ad3b4/api/api_full.go#L119 ? j/k

Reply:
:bike: :derelict_house: 

I wonder what that method does if you call it with chain head :thinking:

Reply:
can I have a :heavy_check_mark: now, please